Add the general convolution operation to extensions

DarrenZhang01 opened the pull request:
TF XLA version:
https://www.tensorflow.org/xla/operation_semantics?hl=en#conv_convolution;
JAX version: https://jax.readthedocs.io/en/stable/_autosummary/jax.lax.conv_general_dilated.html

DarrenZhang01 commented:
tests coming soon.

akshaym commented:
Hey @DarrenZhang01, let me know if I can take a look at this. If there are bugs lets just document them for now and add the code anyway? WDYT?

DarrenZhang01 commented:
> Hey @DarrenZhang01, let me know if I can take a look at this. If there are bugs lets just document them for now and add the code anyway? WDYT?

Thanks very much for the support, Akshay! It is ready for being reviewed.

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+   in the 3D case, if lhs_dilation = 2, then convert it to [2, 2, 2]
+                   if lhs_dilation = (2, 2, 2), convert it also to [2, 2, 2]
+  """
+  strides = [window_strides] * dim if isinstance(window_strides, int) else \
```
Where is dim coming from? 

Alternatively you can define an inner function and just use it on each:
```suggestion
def _as_list_of_size(item, size):
  if item is None:
    return None
  return [item] * size if isinstance(item, int) else list(item)
  
  return _as_list_of_size(window_strides), _as_list_of_size(lhs_dilation), _as_list_of_size(rhs_dilation)
```

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+                         batch_group_count=1, precision=None):
+  """ A general conv API that integrates normal conv, deconvolution,
+  dilated convolution, etc."""
+  dim = None
```
nit: can remove this line?

DarrenZhang01 replied:
I will send this as parameters for the helper function.

akshaym replied:
Setting dim = None is still not required though right? Line 633 below sets dim anyway?

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+  dim = None
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+  if lhs_spec != out_spec:
+    raise TypeError("Current implementation requires the `data_format` of the "
```
These and the below all feel like ValueErrors to me - not TypeErrors.

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+  rhs = np.moveaxis(rhs, (dim_maps['O'], dim_maps['I']), (dim + 1, dim))
+  spatial_dim_maps = {1: 'W', 2: "HW", 3: "DHW"}
+  data_format = 'N' + spatial_dim_maps[dim] + 'C'
+  tf_nn_APIs = {1: [nn.conv1d, nn.conv1d_transpose],
```
Lets make this a module-level constant instead?

DarrenZhang01 replied:
Hi, Akshay, could you elaborate on how to make this "module-level constant"?

DarrenZhang01 replied:
OK, I will do something similar to `_int_dtypes`.

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+  else:
+    output = tf_nn_APIs[dim][1](lhs, rhs, tf.constant(output_shape), strides,
+                                padding, data_format, lhs_dilation)
+  output = np.moveaxis(output, (0, dim + 1), (dim_maps['N'], dim_maps['C']))
```
I think this and below should be using tf_np.moveaxis and tf_np.asarray. 

It'd be good to document the input output expectations. Perhaps similar to dot_general, the name for this function should be "tf_conv_general_dilated", so inputs and outputs are expected to be tf.Tensors, and the tf_np.asarray should not be required.

akshaym commented on trax/tf_numpy/extensions/extensions_test.py:
```diff
@@ -421,6 +422,53 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
+    rhs_tf = tf_np.transpose(tf_np.ones(rhs_shape), rhs_perm)
+
+    lhs_jax = jnp.transpose(jnp.ones(lhs_shape), lhs_perm)
+    rhs_jax = jnp.transpose(jnp.ones(rhs_shape), rhs_perm)
```
Instead of these, don't conv_general_dilated from JAX and TF both accept numpy arrays?

DarrenZhang01 replied:
Yeah, I think vanilla NumPy arrays should not affect function execution here.

akshaym commented on trax/tf_numpy/extensions/extensions_test.py:
```diff
@@ -421,6 +422,53 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
+                             padding, lhs_dilation, rhs_dilation,
+                             feature_group_count, batch_group_count,
+                             dimension_numbers, perms):
+    tf.print("dimension_numbers: {}".format(dimension_numbers), output_stream=sys.stdout)
```
Remove debug print?

akshaym commented on trax/tf_numpy/extensions/extensions_test.py:
```diff
@@ -421,6 +422,53 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
+    jax_conv = jax.lax.conv_general_dilated(lhs_jax, rhs_jax, strides, padding, lhs_dilation,
+      rhs_dilation, dimension_numbers, feature_group_count, batch_group_count)
+
+    tf_conv = lax.conv_general_dilated(lhs_tf, rhs_tf, strides, padding, jax_conv.shape, lhs_dilation,
```
extensions.conv_general_dilated?

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+
+
+# TODO (Zhibo Zhang): Run pylint and complement the docstring.
+# TOTO (Zhibo Zhang): Expand the test cases of general convolution and revise
```
nit spelling TOTO -> TODO

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+                2: [nn.conv2d, nn.conv2d_transpose],
+                3: [nn.conv3d, nn.conv3d_transpose]}
+
+  output = None
```
nit: can remove this line?

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +565,93 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
   return tf.einsum(equation, lhs, rhs)
 
 
+# TODO (Zhibo Zhang): Run pylint and complement the docstring.
```
Lets do both pylint/docstring changes for both functions this and below before submission. 

Also, I think its better to put the TODOs under your github username instead of your full name (so "DarrenZhang01" instead of "Zhibo Zhang").

DarrenZhang01 commented:
> You had mentioned a bug - is that still around? Or is that fixed?

So the original bug is that the transpose convolution was not supported due to the lack of output shape as I mentioned in our last Friday's meeting. Now I added an output shape evaluation helper. 

Besides, I used the test cases from JAX general convolution. Those cases are really broad and TF general conv currently do not cover them all, so I added a TODO to expand the test cases later on.

DarrenZhang01 commented:
Thanks for the review, Akshay!

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -37,6 +38,9 @@
     tf.int64, tf.int32, tf.int16, tf.int8, tf.uint8, tf.uint16, tf.uint32,
     tf.uint64
 ]
+_tf_nn_APIs = {1: [nn.conv1d, nn.conv1d_transpose],
```
nit just use tf.nn.conv1d etc and remove the additional import.

DarrenZhang01 commented:
> Thanks for the review, Akshay!

Now it is ready for review again. @akshaym

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+    feature_group_count:  integer, default 1.
+    batch_group_count: integer, default 1.
+    precision: Optional. Either None, which means the default precision for the
+               backend, or a Precision enum value.
```
Can you add a Returns section specifying that a tf_np.ndarray is returned?

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+    dimension_numbers: either None, a ConvDimensionNumbers object, or a 3-tuple
+                       (lhs_spec, rhs_spec, out_spec), where each element is a
+                       string of length n+2.
+    feature_group_count:  integer, default 1.
```
Add comments for things that are currently not supported. 
For example:
```suggestion
    feature_group_count:  integer, default 1. Changing this is currently not supported. 
```

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+                       .format(lhs_dilation, rhs_dilation))
+  if padding not in ["SAME", "VALID"]:
+    raise ValueError("Current implementation requires the padding parameter"
+                     "to be either 'VALID' or 'SAME', but got: ", padding)
```
Lets also check that feature_group_count, batch_group_count and precision are equal to their defaults. If they are not, we should raise a NotImplementedError.

akshaym commented on trax/tf_numpy/extensions/extensions_test.py:
```diff
@@ -421,6 +424,56 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
+    lhs = np.transpose(np.ones(lhs_shape), lhs_perm)
+    rhs = np.transpose(np.ones(rhs_shape), rhs_perm)
+
+    jax_conv = jax.lax.conv_general_dilated(lhs, rhs, strides, padding,
```
lax.conv_general_dilated and remove the import jax statement from above.

akshaym commented on trax/tf_numpy/extensions/extensions_test.py:
```diff
@@ -421,6 +424,56 @@ def test_tf_dot_general(self, lhs_np, rhs_np, dims):
+                                            batch_group_count)
+
+    tf_conv = extensions.tf_conv_general_dilated(lhs, rhs, strides,
+                                                 padding, jax_conv.shape,
```
Can you make this just be a None instead of jax_conv.shape? Looks like we are not testing lhs_dilation in any case?

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +568,113 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+    padding: either the string ‘SAME’, the string ‘VALID’, or a sequence of n
+             (low, high) integer pairs that give the padding to apply before and
+             after each spatial dimension.
+    output_shape: the output shape of the convolution.
```
Point out that its only required for transposed convolution?

DarrenZhang01 commented:
Thanks a lot, Akshay!

akshaym commented on trax/tf_numpy/extensions/extensions.py:
```diff
@@ -564,6 +567,125 @@ def tf_dot_general(lhs, rhs, dimension_numbers):
+                              "other than 1 is currently not supported, but"
+                              " got feature_group_count: {}, batch_group_count"
+                              ": {}".format(feature_group_count,
+                                            batch_group_count))
```
Check precision also?
